far/global.hpp

```diff
@@ -42,14 +42,58 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "platform.hpp"
 
 // Common:
+#include "common/bytes_view.hpp"
 #include "common/preprocessor.hpp"
 
 // External:
 
 //----------------------------------------------------------------------------
 
-enum class lng;
-enum class search_case_fold;
+struct global_search_replace_params
+{
+	NONCOPYABLE(global_search_replace_params);
+	NONMOVABLE(global_search_replace_params);
+
+	global_search_replace_params() = default;
+
+						// It is client's responsibility to maintain invariants:
+	string SearchStr;   // If Hex, should be BytesToString(SearchBytes, CodePage)
+	bytes SearchBytes;  // If not Hex, should be empty
+	string ReplaceStr;
+	bool Hex{};
+	bool CaseSensitive{};
+	bool WholeWords{};
+	bool Regex{};
+	bool Fuzzy{};
+	bool PreserveStyle{};
+};
+
+template <typename client_params_t>
+void operator<<(global_search_replace_params& GlobalParams, const client_params_t& ClientParams)
+{
+	const auto SaveParam{ [](auto& GlobalParam, const auto& ClientParam)
+	{
+		if constexpr (requires { ClientParam.has_value(); })
+		{
+			if (ClientParam.has_value())
+				GlobalParam = ClientParam.value();
+		}
+		else
+		{
+			GlobalParam = ClientParam;
+		}
+	} };
+
+	if constexpr (requires { ClientParams.SearchStr; })     SaveParam(GlobalParams.SearchStr, ClientParams.SearchStr);
+	if constexpr (requires { ClientParams.SearchBytes; })   SaveParam(GlobalParams.SearchBytes, ClientParams.SearchBytes);
+	if constexpr (requires { ClientParams.ReplaceStr; })    SaveParam(GlobalParams.ReplaceStr, ClientParams.ReplaceStr);
+	if constexpr (requires { ClientParams.Hex; })           SaveParam(GlobalParams.Hex, ClientParams.Hex);
+	if constexpr (requires { ClientParams.CaseSensitive; }) SaveParam(GlobalParams.CaseSensitive, ClientParams.CaseSensitive);
+	if constexpr (requires { ClientParams.WholeWords; })    SaveParam(GlobalParams.WholeWords, ClientParams.WholeWords);
+	if constexpr (requires { ClientParams.Regex; })         SaveParam(GlobalParams.Regex, ClientParams.Regex);
+	if constexpr (requires { ClientParams.Fuzzy; })         SaveParam(GlobalParams.Fuzzy, ClientParams.Fuzzy);
+	if constexpr (requires { ClientParams.PreserveStyle; }) SaveParam(GlobalParams.PreserveStyle, ClientParams.PreserveStyle);
```

Comment:
This is an interesting pattern, but it looks rather fragile: if someone renames a parameter in ClientParams (to be more descriptive or whatever), this place can be easily overlooked because the compiler won't complain at all.

Maybe just make the struct copyable & assignable and let the clients have their copies?

If we decide to proceed with the suggested above hex logic encapsulation,
the data-only part of the structure can be extracted into a base structure and shared with the clients.

Reply:
There are three places where we need these parameters in somewhat different capacities:

1. The clients: View, Edit, Help, Find File. Each of them needs different subset of parameters.
2. The Search / Relace dialog. It needs to have all parameters and deal with either combination of them to be optional. Currently this is modelled with `std::optional`. It can also be done with templates.
3. The Global. It heeds to have all parameters, but the save operation must not override those which do not exist in the client, so we need some way to indicate which parameters exist in this particular client, and which should be ignored (to preserve the values saved by the clients who actually have them).

So, the client's parameter subset serves as a filter between the dialog and the global. The proper design is to have three distinct types: full set of non-optional parameters in global, full set of optional params in the dialog, and proper subset of non-optional params in each of the clients. Then, we need two mappings, between client's and dialog's, and between client's and global's. The mappings should be smart enough to leave optional params in dialog's empty if there are no client's counterparts, and to not reset global's params with non-existent client's params.

I actually prototyped this approach. There are two disadvantages. First of all, it is unwieldy, lots of repeatable code even with templated everything. Second, it obviously involves unnecessary copying of values at run time.

The other extreme implementation is to use `SearchReplaceDlgParams` (properly renamed) in all three places. There are some issues here as well.
1. The use of optional in global seems to be utterly wrong. All global params always have some values (even if `false` immediately after start).
2. The use of optional in Find File is completely redundant. and will only clutter the code.
3. The clients still need to indicate which parameters are applicable, so somewhere each of them needs to list all parameters anyway.

I prototyped this approach as well. It does not feel elegant either. Let me try it once again, and let's see.

Reply:
Another approach is perhaps just setting the required fields explicitly?
```C++
SearchReplaceParamsCommon.Regex = LastSearchDlgParams.Regex;
SearchReplaceParamsCommon.CaseSensitive = LastSearchDlgParams.CaseSensitive;
...
```
A bit more typing, but probably more robust than smart templates. And it's not like we need to do it 100 times, only 4.